src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,18 +7166,88 @@ bool canFollowNullableType()
                     case ParseTypeMode.AfterIs:
                     case ParseTypeMode.DefinitePattern:
                     case ParseTypeMode.AsExpression:
-                        // These contexts might be a type that is at the end of an expression. In these contexts we only
-                        // permit the nullable qualifier if it is followed by a token that could not start an
-                        // expression, because for backward compatibility we want to consider a `?` token as part of the
-                        // `?:` operator if possible.
+                        // We are currently after `?` token after a nullable type pattern and need to decide how to
+                        // parse what we see next.  In the case of an identifier (e.g. `x ? a` there are two ways we can
+                        // see things
                         //
-                        // Similarly, if we have `T?[]` we do want to treat that as an array of nullables (following
-                        // existing parsing), not a conditional that returns a list.
+                        // 1. As a start of conditional expression, e.g. `var a = obj is string ? a : b`
+                        // 2. As a designation of a nullable-typed pattern, e.g. `if (obj is string? str)`
+                        //
+                        // Since nullable types (no matter reference of value types) are not valid in patterns by
+                        // default we are biased towards the first option and consider case 2 only for error recovery
+                        // purposes (if we parse here as nullable type pattern an error will be reported during
+                        // binding). This condition checks for simple cases, where we better use option 2 and parse a
+                        // nullable-typed pattern
+                        if (IsTrueIdentifier(this.CurrentToken))
+                        {
+                            // If we're in a cast statement/expression arm, then we have:
+                            //
+                            // `switch { case X ? y` or
+                            //
+                            //  e switch { X ? y `
```

Comment:
Similarly, this assumption is closer to correct, but [isn't always](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs7sgDsBnDHWAQIuIXQC8dDFACuMANwcuvAXWw9B2UpLqkllLtwj8tm9cV3EDR2gxR0AsgAoNWmgEo2yo5wD6utj0fADu2BhgABbehr5xRs7AdAD86jogFl4SAHx0ND6+AL42XIVUhUA===):

```cs
        _ = i3 switch {
                (b ? i1 : i2) => 3
        };
```

You do need the parentheses to force parse in this manner, but I'm concerned that we'll still end up with bad behavior here.

Reply:
Given that this requires parentheses, it wasn't affected even with the wrong assumption. Added a test for this case